source/common/upstream/upstream_impl.cc

```diff
@@ -102,6 +109,21 @@ HostImpl::createConnection(Event::Dispatcher& dispatcher, const ClusterInfo& clu
   } else {
     cluster_options = options;
   }
+  if (cluster.features() & ClusterInfo::Features::USE_TCP_KEEPALIVE) {
+    if (!cluster_options) {
+      cluster_options = std::make_shared<Network::ConnectionSocket::Options>();
+      if (options) {
+        *cluster_options = *options;
+      }
+    }
+    cluster_options->emplace_back(new Network::TcpKeepaliveOptionImpl(
+        PROTOBUF_GET_WRAPPED_OR_DEFAULT(cluster.tcpKeepaliveSettings(), keepalive_probes,
```

Comment:
This is horrible and I shouldn't have done it this way. I'll move it out to a struct that'll be populated at construction time.